### PR TITLE
Add Multipart Upload Handlers

### PR DESCRIPTION
**Pull request by gabrielfu**
## Related Issues/PRs

 #9133 #9467

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -1917,6 +1921,117 @@ def _delete_artifact_mlflow_artifacts(artifact_path):
...
+    path = request_message.path
+    num_parts = request_message.num_parts
+
+    artifact_repo = _get_artifact_repo_mlflow_artifacts()
+    if not isinstance(artifact_repo, MultipartUploadMixin):
+        raise MlflowException(
+            "Multipart upload is not supported for artifact repository "
+            f"{artifact_repo.__class__.__name__}",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
```
Comment (gabrielfu) [inline mlflow/server/handlers.py:1949]: i tried refactoring this out, but intellisense won't be able to detect the object type

File: mlflow/entities/multipart_upload.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    url: str
+    part_number: int
+    headers: Dict[str, Any]
+
+    def to_proto(self):
+        credential = ProtoMultipartUploadCredential()
+        credential.url = self.url
+        credential.part_number = self.part_number
+        credential.headers.update(self.headers)
+        return credential
```
Comment (harupy) [inline mlflow/entities/multipart_upload.py:34]: /ai This data class represents a chunk in a multipart upload. `url` is a pre-signed URL and `headers` is headers used when uploading the chunk. How would you name this class?

File: mlflow/server/handlers.py (hunk 1 above)
Comment (harupy) [reply mlflow/server/handlers.py:1950]: > intellisense won't be able to detect the object type

Can you elaborate?

File: mlflow/store/artifact/artifact_repo.py [hunk 3]
```diff
@@ -244,6 +246,58 @@ def max_workers(self) -> int:
         return min(num_cpus * _NUM_MAX_THREADS_PER_CPU, _NUM_MAX_THREADS)
 
 
+class MultipartUploadMixin(ABC):
+    @abstractmethod
+    def create_multipart_upload(
+        self, local_file: str, num_parts: int, artifact_path=None
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:252]:
```suggestion
        self, local_file: str, num_parts: int, artifact_path: Optional[str]=None
```

File: mlflow/store/artifact/artifact_repo.py [hunk 4]
```diff
@@ -244,6 +246,58 @@ def max_workers(self) -> int:
...
+                              artifact.
+        """
+        pass
+
+    @abstractmethod
+    def complete_multipart_upload(
+        self,
+        local_file: str,
+        upload_id: str,
+        parts: List,
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:269]:
```suggestion
        parts: List[...],
```

Can we add the element type?

File: mlflow/store/artifact/artifact_repo.py [hunk 5]
```diff
@@ -244,6 +246,58 @@ def max_workers(self) -> int:
...
+                              artifact.
+        """
+        pass
+
+    @abstractmethod
+    def abort_multipart_upload(
+        self,
+        local_file: str,
+        upload_id: str,
+        artifact_path=None,
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:288]:
```suggestion
        artifact_path: Optional[str]=None,
```

File: mlflow/server/handlers.py [hunk 6]
```diff
@@ -1917,6 +1922,117 @@ def _delete_artifact_mlflow_artifacts(artifact_path):
...
+    path = request_message.path
+    upload_id = request_message.upload_id
+    parts = [MultipartUploadPart.from_proto(part) for part in request_message.parts]
+
+    artifact_repo = _get_artifact_repo_mlflow_artifacts()
+    if not isinstance(artifact_repo, MultipartUploadMixin):
+        raise MlflowException(
+            "Multipart upload is not supported for artifact repository "
+            f"{artifact_repo.__class__.__name__}",
+            error_code=INVALID_PARAMETER_VALUE,
```
Comment (harupy) [inline mlflow/server/handlers.py:1989]:
```suggestion
            error_code=BAD_REQUEST,
```

`BAD_REQUEST` seems more appropriate?

File: mlflow/server/handlers.py (hunk 1 above)
Comment (gabrielfu) [reply mlflow/server/handlers.py]: don't worry about it, I extracted it out to a method

File: mlflow/server/handlers.py (hunk 6 above)
Comment (gabrielfu) [reply mlflow/server/handlers.py]: good point

**Comment by gabrielfu** [comment]
@mlflow-automation autoformat

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**